internal/scheduler/publisher.go

```diff
@@ -2,20 +2,193 @@ package scheduler
 
 import (
 	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
 
+	"github.com/apache/pulsar-client-go/pulsar"
+	"github.com/gogo/protobuf/proto"
 	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
 
+	"github.com/G-Research/armada/internal/common/eventutil"
 	"github.com/G-Research/armada/pkg/armadaevents"
 )
 
+const explicitPartitionKey = "armada_pulsar_partition"
+
 // Publisher is an interface to be implemented by structs that handle publishing messages to pulsar
 type Publisher interface {
 	// PublishMessages will publish the supplied messages. A LeaderToken is provided and the
-	// implementor may decide whether or not to publish based on the status of this token
+	// implementor may decide whether to publish based on the status of this token
 	PublishMessages(ctx context.Context, events []*armadaevents.EventSequence, token LeaderToken) error
 
 	// PublishMarkers publishes a single marker message for each Pulsar partition.  Each marker
 	// massage contains the supplied group id, which allows nall marker messages for a given call
 	// to be identified.  The uint32 returned is the number of messages published
 	PublishMarkers(ctx context.Context, groupId uuid.UUID) (uint32, error)
 }
+
+// PulsarPublisher is the default implementation of Publisher
+type PulsarPublisher struct {
```

Comment:
Is this the right abstraction? I'd prefer leader election and pulsar publishing be separate abstractions, i.e., 1) check if leader and 2) publish. Imo this abstraction is too specialised.